File: drone_gcs/python_service/observability/metrics.py
"""Prometheus metrics for the Python MAVLink service (Phase 5A).

A dedicated CollectorRegistry (not the global default) holds the service's golden
signals plus the telemetry-publisher internals ops needs to watch the F5 ceiling.
Using an explicit registry keeps tests isolated (no cross-test metric bleed) and
lets the app expose exactly this set at GET /metrics.
"""
from __future__ import annotations

from prometheus_client import CollectorRegistry, Counter, Gauge, Histogram, generate_latest, CONTENT_TYPE_LATEST


class Metrics:
    def __init__(self, registry: CollectorRegistry | None = None) -> None:
        self.registry = registry or CollectorRegistry()

        self.http_requests = Counter(
            "http_requests_total", "HTTP requests handled",
            ["method", "route", "status"], registry=self.registry,
        )
        self.http_duration = Histogram(
            "http_request_duration_seconds", "HTTP request duration (s)",
            ["method", "route"],
            buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5),
            registry=self.registry,
        )
        self.telemetry_frames_published = Counter(
            "telemetry_frames_published_total", "ZMQ telemetry frames published",
            ["drone_id", "type"], registry=self.registry,
        )
        self.telemetry_publish_errors = Counter(
            "telemetry_publish_errors_total", "ZMQ publish send errors",
            registry=self.registry,
        )
        self.telemetry_tick_duration = Histogram(
            "telemetry_tick_duration_seconds", "Duration of one publish_loop tick (s)",
            buckets=(0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5),
            registry=self.registry,
        )
        self.telemetry_ticks = Counter(
            "telemetry_ticks_total", "publish_loop ticks executed",
            registry=self.registry,
        )
        self.telemetry_tick_failures = Counter(
            "telemetry_tick_failures_total", "publish_loop ticks that raised",
            registry=self.registry,
        )
        self.fleet_drones = Gauge(
            "fleet_drones", "Registered drone sessions in this worker",
            registry=self.registry,
        )
        self.fleet_drones_online = Gauge(
            "fleet_drones_online", "Drone sessions currently online",
            registry=self.registry,
        )
        self.drone_last_seen_age = Gauge(
            "drone_last_seen_age_seconds", "Seconds since last message per drone",
            ["drone_id"], registry=self.registry,
        )
        self.zmq_publisher_up = Gauge(
            "zmq_publisher_up", "1 when the ZMQ publisher socket is bound and running",
            registry=self.registry,
        )

    def expose(self) -> bytes:
        return generate_latest(self.registry)

    @property
    def content_type(self) -> str:
        return CONTENT_TYPE_LATEST


_metrics: Metrics | None = None


def get_metrics() -> Metrics:
    """Process-wide Metrics singleton (created lazily)."""
    global _metrics
    if _metrics is None:
        _metrics = Metrics()
    return _metrics

File: drone_gcs/node_api/lib/metrics.js
/**
 * Prometheus-compatible metrics for the Node API gateway (Phase 5A).
 *
 * A single Registry holds the golden-signal metrics for the gateway plus the
 * telemetry/ZMQ/WS internals the fleet ops team needs to see the F5 ceiling
 * approaching. Exposed at GET /metrics in text exposition format.
 */
'use strict';

const client = require('prom-client');

function createMetrics({ serviceName = 'drone-gcs-node-api', collectDefault = true } = {}) {
  const registry = new client.Registry();
  registry.setDefaultLabels({ service: serviceName });
  if (collectDefault) {
    client.collectDefaultMetrics({ register: registry });
  }

  const httpRequests = new client.Counter({
    name: 'http_requests_total',
    help: 'Total HTTP requests handled by the gateway',
    labelNames: ['method', 'route', 'status'],
    registers: [registry],
  });

  const httpDuration = new client.Histogram({
    name: 'http_request_duration_seconds',
    help: 'HTTP request duration in seconds',
    labelNames: ['method', 'route', 'status'],
    buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5],
    registers: [registry],
  });

  const wsClients = new client.Gauge({
    name: 'ws_clients',
    help: 'Currently connected telemetry WebSocket clients',
    registers: [registry],
  });

  const wsMessagesSent = new client.Counter({
    name: 'ws_messages_sent_total',
    help: 'Telemetry frames sent to WebSocket clients',
    registers: [registry],
  });

  const wsSendDropped = new client.Counter({
    name: 'ws_send_dropped_total',
    help: 'Telemetry frames dropped due to a slow-client send-queue overflow',
    labelNames: ['reason'],
    registers: [registry],
  });

  const wsClientsDisconnected = new client.Counter({
    name: 'ws_clients_disconnected_total',
    help: 'WebSocket clients force-closed by the server',
    labelNames: ['reason'],
    registers: [registry],
  });

  const zmqFramesReceived = new client.Counter({
    name: 'zmq_frames_received_total',
    help: 'Frames received from the Python telemetry publisher over ZMQ',
    registers: [registry],
  });

  const zmqReconnects = new client.Counter({
    name: 'zmq_reconnects_total',
    help: 'ZMQ subscriber reconnect attempts after an error or stream end',
    registers: [registry],
  });

  const zmqConnected = new client.Gauge({
    name: 'zmq_connected',
    help: '1 when the ZMQ subscriber is connected and looping, else 0',
    registers: [registry],
  });

  const zmqLastFrameAge = new client.Gauge({
    name: 'zmq_last_frame_age_seconds',
    help: 'Seconds since the last ZMQ telemetry frame was received',
    registers: [registry],
  });

  const telemetryFramesProcessed = new client.Counter({
    name: 'telemetry_frames_processed_total',
    help: 'Telemetry frames processed by the enrichment engine',
    labelNames: ['drone_id'],
    registers: [registry],
  });

  const droneLastSeenAge = new client.Gauge({
    name: 'drone_last_seen_age_seconds',
    help: 'Seconds since a frame was last processed for a given drone_id',
    labelNames: ['drone_id'],
    registers: [registry],
  });

  return {
    client,
    registry,
    httpRequests,
    httpDuration,
    wsClients,
    wsMessagesSent,
    wsSendDropped,
    wsClientsDisconnected,
    zmqFramesReceived,
    zmqReconnects,
    zmqConnected,
    zmqLastFrameAge,
    telemetryFramesProcessed,
    droneLastSeenAge,
    async expose() {
      return registry.metrics();
    },
    contentType: registry.contentType,
  };
}

module.exports = { createMetrics };
